src/elf.rs

```diff
@@ -660,6 +689,46 @@ impl<'a> Elf<'a> {
                 8 => (reloc::R_X86_64_64, addend),
                 _ => panic!("unsupported relocation {:?}", l),
             },
+            Reloc::Data { addend } => {
+                match *l.to.decl {
+                    Decl::Defined(DefinedDecl::Data { .. })
+                    | Decl::Import(ImportKind::Data { .. }) => {}
+                    _ => panic!("unsupported relocation {:?}", l.to.decl),
+                };
+
+                match *l.from.decl {
+                    Decl::Defined(DefinedDecl::Function { .. }) => match *l.to.decl {
+                        Decl::Defined(DefinedDecl::Data { .. }) => {
+                            // We're referencing a symbol that we've directly
+                            // defined, so we're not going through the GOT.
+                            // This means that the addend will be applied
+                            // directly to the address of the target symbol,
+                            // which can be used to reference a specific location
+                            // within the target (e.g. a particular element of an array)
+                            (reloc::R_X86_64_PC32, X64_IMM_OFFSET + addend)
+                        }
+                        Decl::Import(ImportKind::Data) => {
+                            // If we're linking to an imported symbol,
+                            // we're creating a GOT-relative relocation.
+                            // Having a user-supplied addend doesn't make sense here,
+                            // as it would cause us to move within the GOT itself -
+                            // not relative to the address stored in the GOT entry.
+                            assert!(addend == 0, "Addend must be 0 for reloc {:?}", l);
```

Comment:
so the way we've been doing this, or at least I thought, perhaps things have changed:
1. artifact is the only real "entry" to the various backends.
2. consequently must user level errors are attempted to be caught during artifact construction, and a Result is returned.
3. asserts in the backends are to ensure that contracts between the artifact front end and the backend are enforced at the artifact front end via result. if there is some kind of user introduced error that is backend specific, but not enforced by artifact (because it can't, for example), then we should return a result, instead of asserting.

So I bring this all up because this assert looks like its enforcing user inputs to have certain properties, which means it should be a returning a Result.
That being said, I can't see on this git review if this function returns a Result, so maybe you didn't feel like changing the return type, etc., which is reasonable, but then we should revisit (at some other PR perhaps) this functions return type if user errors are being asserted against.

Reply:
It doesn't return a Result (FYI github lets you expand the review code to see this).